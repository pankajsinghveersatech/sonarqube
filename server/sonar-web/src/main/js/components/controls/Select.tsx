/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import ReactSelectClass, {
  ReactAsyncSelectProps,
  ReactCreatableSelectProps,
  ReactSelectProps
} from 'react-select';
import { lazyLoadComponent } from '../lazyLoadComponent';
import { ClearButton } from './buttons';
import './Select.css';

const ReactSelectLib = import('react-select');
const ReactSelect = lazyLoadComponent(() => ReactSelectLib);
const ReactCreatable = lazyLoadComponent(() =>
  ReactSelectLib.then(lib => ({ default: lib.Creatable }))
);
const ReactAsync = lazyLoadComponent(() => ReactSelectLib.then(lib => ({ default: lib.Async })));

function renderInput() {
  return <ClearButton className="button-tiny spacer-left text-middle" iconProps={{ size: 12 }} />;
}

export interface WithInnerRef {
  innerRef?: React.Ref<ReactSelectClass<unknown>>;
}

export default function Select({ innerRef, ...props }: WithInnerRef & ReactSelectProps) {
  // hide the "x" icon when select is empty
  const clearable = props.clearable ? Boolean(props.value) : false;
  return (
    <ReactSelect {...props} clearable={clearable} clearRenderer={renderInput} ref={innerRef} />
  );
}

export function Creatable(props: ReactCreatableSelectProps) {
  return <ReactCreatable {...props} clearRenderer={renderInput} />;
}

export function AsyncSelect(props: ReactAsyncSelectProps & WithInnerRef) {
  return <ReactAsync {...props} />;
}
