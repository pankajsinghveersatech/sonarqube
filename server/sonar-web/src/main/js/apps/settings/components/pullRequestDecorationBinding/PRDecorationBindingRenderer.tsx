/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { FormattedMessage } from 'react-intl';
import { Link } from 'react-router';
import { Button, SubmitButton } from '../../../../components/controls/buttons';
import Select from '../../../../components/controls/Select';
import AlertSuccessIcon from '../../../../components/icons/AlertSuccessIcon';
import { Alert } from '../../../../components/ui/Alert';
import DeferredSpinner from '../../../../components/ui/DeferredSpinner';
import MandatoryFieldMarker from '../../../../components/ui/MandatoryFieldMarker';
import MandatoryFieldsExplanation from '../../../../components/ui/MandatoryFieldsExplanation';
import { translate } from '../../../../helpers/l10n';
import { getGlobalSettingsUrl } from '../../../../helpers/urls';
import {
  AlmSettingsInstance,
  ProjectAlmBindingConfigurationErrors,
  ProjectAlmBindingConfigurationErrorScope,
  ProjectAlmBindingResponse
} from '../../../../types/alm-settings';
import { ALM_INTEGRATION } from '../AdditionalCategoryKeys';
import AlmSpecificForm from './AlmSpecificForm';

export interface PRDecorationBindingRendererProps {
  formData: T.Omit<ProjectAlmBindingResponse, 'alm'>;
  instances: AlmSettingsInstance[];
  isChanged: boolean;
  isConfigured: boolean;
  isValid: boolean;
  loading: boolean;
  onFieldChange: (id: keyof ProjectAlmBindingResponse, value: string | boolean) => void;
  onReset: () => void;
  onSubmit: () => void;
  updating: boolean;
  successfullyUpdated: boolean;
  monorepoEnabled: boolean;
  onCheckConfiguration: () => void;
  checkingConfiguration: boolean;
  configurationErrors?: ProjectAlmBindingConfigurationErrors;
  isSysAdmin: boolean;
}

function optionRenderer(instance: AlmSettingsInstance) {
  return instance.url ? (
    <>
      <span>{instance.key} â€” </span>
      <span className="text-muted">{instance.url}</span>
    </>
  ) : (
    <span>{instance.key}</span>
  );
}

export default function PRDecorationBindingRenderer(props: PRDecorationBindingRendererProps) {
  const {
    formData,
    instances,
    isChanged,
    isConfigured,
    isValid,
    loading,
    updating,
    successfullyUpdated,
    monorepoEnabled,
    checkingConfiguration,
    configurationErrors,
    isSysAdmin
  } = props;

  if (loading) {
    return <DeferredSpinner />;
  }

  if (instances.length < 1) {
    return (
      <div>
        <Alert className="spacer-top huge-spacer-bottom" variant="info">
          {isSysAdmin ? (
            <FormattedMessage
              defaultMessage={translate('settings.pr_decoration.binding.no_bindings.admin')}
              id="settings.pr_decoration.binding.no_bindings.admin"
              values={{
                link: (
                  <Link to={getGlobalSettingsUrl(ALM_INTEGRATION)}>
                    {translate('settings.pr_decoration.binding.no_bindings.link')}
                  </Link>
                )
              }}
            />
          ) : (
            translate('settings.pr_decoration.binding.no_bindings')
          )}
        </Alert>
      </div>
    );
  }

  const selected = formData.key && instances.find(i => i.key === formData.key);
  const alm = selected && selected.alm;

  return (
    <div>
      <header className="page-header">
        <h1 className="page-title">{translate('settings.pr_decoration.binding.title')}</h1>
      </header>

      <div className="markdown small spacer-top big-spacer-bottom">
        {translate('settings.pr_decoration.binding.description')}
      </div>

      <form
        onSubmit={(event: React.SyntheticEvent<HTMLFormElement>) => {
          event.preventDefault();
          props.onSubmit();
        }}>
        <MandatoryFieldsExplanation className="form-field" />

        <div className="settings-definition big-spacer-bottom">
          <div className="settings-definition-left">
            <label className="h3" htmlFor="name">
              {translate('settings.pr_decoration.binding.form.name')}
              <MandatoryFieldMarker className="spacer-right" />
            </label>
            <div className="markdown small spacer-top">
              {translate('settings.pr_decoration.binding.form.name.help')}
            </div>
          </div>
          <div className="settings-definition-right">
            <Select
              autosize={true}
              className="abs-width-400 big-spacer-top"
              clearable={false}
              id="name"
              menuContainerStyle={{
                maxWidth: '210%' /* Allow double the width of the select */,
                width: 'auto'
              }}
              onChange={(instance: AlmSettingsInstance) => props.onFieldChange('key', instance.key)}
              optionRenderer={optionRenderer}
              options={instances}
              searchable={false}
              value={formData.key}
              valueKey="key"
              valueRenderer={optionRenderer}
            />
          </div>
        </div>

        {alm && (
          <AlmSpecificForm
            alm={alm}
            instances={instances}
            formData={formData}
            onFieldChange={props.onFieldChange}
            monorepoEnabled={monorepoEnabled}
          />
        )}

        <div className="display-flex-center big-spacer-top action-section">
          {isChanged && (
            <SubmitButton className="spacer-right button-success" disabled={updating || !isValid}>
              <span data-test="project-settings__alm-save">{translate('save')}</span>
              <DeferredSpinner className="spacer-left" loading={updating} />
            </SubmitButton>
          )}
          {!updating && successfullyUpdated && (
            <span className="text-success spacer-right">
              <AlertSuccessIcon className="spacer-right" />
              {translate('settings.state.saved')}
            </span>
          )}
          {isConfigured && (
            <>
              <Button className="spacer-right" onClick={props.onReset}>
                <span data-test="project-settings__alm-reset">{translate('reset_verb')}</span>
              </Button>
              {!isChanged && (
                <Button onClick={props.onCheckConfiguration} disabled={checkingConfiguration}>
                  {translate('settings.pr_decoration.binding.check_configuration')}
                  <DeferredSpinner className="spacer-left" loading={checkingConfiguration} />
                </Button>
              )}
            </>
          )}
        </div>
        {!checkingConfiguration && configurationErrors?.errors && (
          <Alert variant="error" display="inline" className="big-spacer-top">
            <p className="spacer-bottom">
              {translate('settings.pr_decoration.binding.check_configuration.failure')}
            </p>
            <ul className="list-styled">
              {configurationErrors.errors.map((error, i) => (
                // eslint-disable-next-line react/no-array-index-key
                <li key={i}>{error.msg}</li>
              ))}
            </ul>
            {configurationErrors.scope === ProjectAlmBindingConfigurationErrorScope.Global && (
              <p>
                {isSysAdmin ? (
                  <FormattedMessage
                    id="settings.pr_decoration.binding.check_configuration.failure.check_global_settings"
                    defaultMessage={translate(
                      'settings.pr_decoration.binding.check_configuration.failure.check_global_settings'
                    )}
                    values={{
                      link: (
                        <Link to={getGlobalSettingsUrl(ALM_INTEGRATION, { alm })}>
                          {translate(
                            'settings.pr_decoration.binding.check_configuration.failure.check_global_settings.link'
                          )}
                        </Link>
                      )
                    }}
                  />
                ) : (
                  translate('settings.pr_decoration.binding.check_configuration.contact_admin')
                )}
              </p>
            )}
          </Alert>
        )}
        {isConfigured && !isChanged && !checkingConfiguration && !configurationErrors && (
          <Alert variant="success" display="inline" className="big-spacer-top">
            {translate('settings.pr_decoration.binding.check_configuration.success')}
          </Alert>
        )}
      </form>
    </div>
  );
}
