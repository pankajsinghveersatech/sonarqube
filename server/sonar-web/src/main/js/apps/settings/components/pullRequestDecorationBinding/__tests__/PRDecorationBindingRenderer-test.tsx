/*
 * SonarQube
 * Copyright (C) 2009-2021 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import Select from '../../../../../components/controls/Select';
import { waitAndUpdate } from '../../../../../helpers/testUtils';
import {
  AlmKeys,
  AlmSettingsInstance,
  ProjectAlmBindingConfigurationErrors,
  ProjectAlmBindingConfigurationErrorScope
} from '../../../../../types/alm-settings';
import PRDecorationBindingRenderer, {
  PRDecorationBindingRendererProps
} from '../PRDecorationBindingRenderer';

const urls = ['http://github.enterprise.com', 'http://bbs.enterprise.com'];
const instances: AlmSettingsInstance[] = [
  {
    alm: AlmKeys.GitHub,
    key: 'i1',
    url: urls[0]
  },
  {
    alm: AlmKeys.GitHub,
    key: 'i2',
    url: urls[0]
  },
  {
    alm: AlmKeys.BitbucketServer,
    key: 'i3',
    url: urls[1]
  },
  {
    alm: AlmKeys.Azure,
    key: 'i4'
  }
];
const configurationErrors: ProjectAlmBindingConfigurationErrors = {
  scope: ProjectAlmBindingConfigurationErrorScope.Global,
  errors: [{ msg: 'Test' }, { msg: 'tesT' }]
};

it.each([
  ['when loading', { loading: true }],
  ['with no ALM instances (admin user)', { isSysAdmin: true }],
  ['with no ALM instances (non-admin user)', {}],
  ['with a single ALM instance', { instances: [instances[0]] }],
  ['with an empty form', { instances }],
  [
    'with a valid and saved form',
    {
      formData: {
        key: 'i1',
        repository: 'account/repo',
        monorepo: false
      },
      isChanged: false,
      isConfigured: true,
      instances
    }
  ],
  [
    'when there are configuration errors (non-admin user)',
    { instances, isConfigured: true, configurationErrors }
  ],
  [
    'when there are configuration errors (admin user)',
    {
      formData: {
        key: 'i1',
        repository: 'account/repo',
        monorepo: false
      },
      instances,
      isConfigured: true,
      configurationErrors,
      isSysAdmin: true
    }
  ],
  [
    'when there are configuration errors (admin user) and error are at PROJECT level',
    {
      instances,
      isConfigured: true,
      configurationErrors: {
        ...configurationErrors,
        scope: ProjectAlmBindingConfigurationErrorScope.Project
      },
      isSysAdmin: true
    }
  ]
])('should render correctly', (name: string, props: PRDecorationBindingRendererProps) => {
  expect(shallowRender(props)).toMatchSnapshot(name);
});

it.each([
  ['updating', { updating: true }],
  ['update is successfull', { successfullyUpdated: true }],
  ['form is valid', { isValid: true }],
  ['configuration is saved', { isConfigured: true }],
  ['configuration check is in progress', { isConfigured: true, checkingConfiguration: true }]
])(
  'should display action section correctly when',
  (name: string, props: PRDecorationBindingRendererProps) => {
    expect(shallowRender({ ...props, instances }).find('.action-section')).toMatchSnapshot(name);
  }
);

it('should render select options correctly', async () => {
  const wrapper = shallowRender({ instances });

  await waitAndUpdate(wrapper);

  const { optionRenderer } = wrapper.find(Select).props();

  expect(optionRenderer!(instances[0])).toMatchSnapshot();

  expect(optionRenderer!(instances[1])).toMatchSnapshot();
});

function shallowRender(props: Partial<PRDecorationBindingRendererProps> = {}) {
  return shallow(
    <PRDecorationBindingRenderer
      formData={{
        key: '',
        repository: '',
        monorepo: false
      }}
      instances={[]}
      isChanged={false}
      isConfigured={false}
      isValid={false}
      loading={false}
      onFieldChange={jest.fn()}
      onReset={jest.fn()}
      onSubmit={jest.fn()}
      updating={false}
      successfullyUpdated={false}
      monorepoEnabled={false}
      checkingConfiguration={false}
      onCheckConfiguration={jest.fn()}
      isSysAdmin={false}
      {...props}
    />
  );
}
