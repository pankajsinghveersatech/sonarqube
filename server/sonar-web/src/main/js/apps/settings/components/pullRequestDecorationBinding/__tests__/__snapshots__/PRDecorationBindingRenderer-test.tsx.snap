// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display action section correctly when: configuration check is in progress 1`] = `
<div
  className="display-flex-center big-spacer-top action-section"
>
  <Button
    className="spacer-right"
    onClick={[MockFunction]}
  >
    <span
      data-test="project-settings__alm-reset"
    >
      reset_verb
    </span>
  </Button>
  <Button
    disabled={true}
    onClick={[MockFunction]}
  >
    settings.pr_decoration.binding.check_configuration
    <DeferredSpinner
      className="spacer-left"
      loading={true}
    />
  </Button>
</div>
`;

exports[`should display action section correctly when: configuration is saved 1`] = `
<div
  className="display-flex-center big-spacer-top action-section"
>
  <Button
    className="spacer-right"
    onClick={[MockFunction]}
  >
    <span
      data-test="project-settings__alm-reset"
    >
      reset_verb
    </span>
  </Button>
  <Button
    disabled={false}
    onClick={[MockFunction]}
  >
    settings.pr_decoration.binding.check_configuration
    <DeferredSpinner
      className="spacer-left"
      loading={false}
    />
  </Button>
</div>
`;

exports[`should display action section correctly when: form is valid 1`] = `
<div
  className="display-flex-center big-spacer-top action-section"
/>
`;

exports[`should display action section correctly when: update is successfull 1`] = `
<div
  className="display-flex-center big-spacer-top action-section"
>
  <span
    className="text-success spacer-right"
  >
    <AlertSuccessIcon
      className="spacer-right"
    />
    settings.state.saved
  </span>
</div>
`;

exports[`should display action section correctly when: updating 1`] = `
<div
  className="display-flex-center big-spacer-top action-section"
/>
`;

exports[`should render correctly: when loading 1`] = `<DeferredSpinner />`;

exports[`should render correctly: when there are configuration errors (admin user) 1`] = `
<div>
  <header
    className="page-header"
  >
    <h1
      className="page-title"
    >
      settings.pr_decoration.binding.title
    </h1>
  </header>
  <div
    className="markdown small spacer-top big-spacer-bottom"
  >
    settings.pr_decoration.binding.description
  </div>
  <form
    onSubmit={[Function]}
  >
    <MandatoryFieldsExplanation
      className="form-field"
    />
    <div
      className="settings-definition big-spacer-bottom"
    >
      <div
        className="settings-definition-left"
      >
        <label
          className="h3"
          htmlFor="name"
        >
          settings.pr_decoration.binding.form.name
          <MandatoryFieldMarker
            className="spacer-right"
          />
        </label>
        <div
          className="markdown small spacer-top"
        >
          settings.pr_decoration.binding.form.name.help
        </div>
      </div>
      <div
        className="settings-definition-right"
      >
        <Select
          autosize={true}
          className="abs-width-400 big-spacer-top"
          clearable={false}
          id="name"
          menuContainerStyle={
            Object {
              "maxWidth": "210%",
              "width": "auto",
            }
          }
          onChange={[Function]}
          optionRenderer={[Function]}
          options={
            Array [
              Object {
                "alm": "github",
                "key": "i1",
                "url": "http://github.enterprise.com",
              },
              Object {
                "alm": "github",
                "key": "i2",
                "url": "http://github.enterprise.com",
              },
              Object {
                "alm": "bitbucket",
                "key": "i3",
                "url": "http://bbs.enterprise.com",
              },
              Object {
                "alm": "azure",
                "key": "i4",
              },
            ]
          }
          searchable={false}
          value="i1"
          valueKey="key"
          valueRenderer={[Function]}
        />
      </div>
    </div>
    <AlmSpecificForm
      alm="github"
      formData={
        Object {
          "key": "i1",
          "monorepo": false,
          "repository": "account/repo",
        }
      }
      instances={
        Array [
          Object {
            "alm": "github",
            "key": "i1",
            "url": "http://github.enterprise.com",
          },
          Object {
            "alm": "github",
            "key": "i2",
            "url": "http://github.enterprise.com",
          },
          Object {
            "alm": "bitbucket",
            "key": "i3",
            "url": "http://bbs.enterprise.com",
          },
          Object {
            "alm": "azure",
            "key": "i4",
          },
        ]
      }
      monorepoEnabled={false}
      onFieldChange={[MockFunction]}
    />
    <div
      className="display-flex-center big-spacer-top action-section"
    >
      <Button
        className="spacer-right"
        onClick={[MockFunction]}
      >
        <span
          data-test="project-settings__alm-reset"
        >
          reset_verb
        </span>
      </Button>
      <Button
        disabled={false}
        onClick={[MockFunction]}
      >
        settings.pr_decoration.binding.check_configuration
        <DeferredSpinner
          className="spacer-left"
          loading={false}
        />
      </Button>
    </div>
    <Alert
      className="big-spacer-top"
      display="inline"
      variant="error"
    >
      <p
        className="spacer-bottom"
      >
        settings.pr_decoration.binding.check_configuration.failure
      </p>
      <ul
        className="list-styled"
      >
        <li
          key="0"
        >
          Test
        </li>
        <li
          key="1"
        >
          tesT
        </li>
      </ul>
      <p>
        <FormattedMessage
          defaultMessage="settings.pr_decoration.binding.check_configuration.failure.check_global_settings"
          id="settings.pr_decoration.binding.check_configuration.failure.check_global_settings"
          values={
            Object {
              "link": <Link
                onlyActiveOnIndex={false}
                style={Object {}}
                to={
                  Object {
                    "pathname": "/admin/settings",
                    "query": Object {
                      "alm": "github",
                      "category": "almintegration",
                    },
                  }
                }
              >
                settings.pr_decoration.binding.check_configuration.failure.check_global_settings.link
              </Link>,
            }
          }
        />
      </p>
    </Alert>
  </form>
</div>
`;

exports[`should render correctly: when there are configuration errors (admin user) and error are at PROJECT level 1`] = `
<div>
  <header
    className="page-header"
  >
    <h1
      className="page-title"
    >
      settings.pr_decoration.binding.title
    </h1>
  </header>
  <div
    className="markdown small spacer-top big-spacer-bottom"
  >
    settings.pr_decoration.binding.description
  </div>
  <form
    onSubmit={[Function]}
  >
    <MandatoryFieldsExplanation
      className="form-field"
    />
    <div
      className="settings-definition big-spacer-bottom"
    >
      <div
        className="settings-definition-left"
      >
        <label
          className="h3"
          htmlFor="name"
        >
          settings.pr_decoration.binding.form.name
          <MandatoryFieldMarker
            className="spacer-right"
          />
        </label>
        <div
          className="markdown small spacer-top"
        >
          settings.pr_decoration.binding.form.name.help
        </div>
      </div>
      <div
        className="settings-definition-right"
      >
        <Select
          autosize={true}
          className="abs-width-400 big-spacer-top"
          clearable={false}
          id="name"
          menuContainerStyle={
            Object {
              "maxWidth": "210%",
              "width": "auto",
            }
          }
          onChange={[Function]}
          optionRenderer={[Function]}
          options={
            Array [
              Object {
                "alm": "github",
                "key": "i1",
                "url": "http://github.enterprise.com",
              },
              Object {
                "alm": "github",
                "key": "i2",
                "url": "http://github.enterprise.com",
              },
              Object {
                "alm": "bitbucket",
                "key": "i3",
                "url": "http://bbs.enterprise.com",
              },
              Object {
                "alm": "azure",
                "key": "i4",
              },
            ]
          }
          searchable={false}
          value=""
          valueKey="key"
          valueRenderer={[Function]}
        />
      </div>
    </div>
    <div
      className="display-flex-center big-spacer-top action-section"
    >
      <Button
        className="spacer-right"
        onClick={[MockFunction]}
      >
        <span
          data-test="project-settings__alm-reset"
        >
          reset_verb
        </span>
      </Button>
      <Button
        disabled={false}
        onClick={[MockFunction]}
      >
        settings.pr_decoration.binding.check_configuration
        <DeferredSpinner
          className="spacer-left"
          loading={false}
        />
      </Button>
    </div>
    <Alert
      className="big-spacer-top"
      display="inline"
      variant="error"
    >
      <p
        className="spacer-bottom"
      >
        settings.pr_decoration.binding.check_configuration.failure
      </p>
      <ul
        className="list-styled"
      >
        <li
          key="0"
        >
          Test
        </li>
        <li
          key="1"
        >
          tesT
        </li>
      </ul>
    </Alert>
  </form>
</div>
`;

exports[`should render correctly: when there are configuration errors (non-admin user) 1`] = `
<div>
  <header
    className="page-header"
  >
    <h1
      className="page-title"
    >
      settings.pr_decoration.binding.title
    </h1>
  </header>
  <div
    className="markdown small spacer-top big-spacer-bottom"
  >
    settings.pr_decoration.binding.description
  </div>
  <form
    onSubmit={[Function]}
  >
    <MandatoryFieldsExplanation
      className="form-field"
    />
    <div
      className="settings-definition big-spacer-bottom"
    >
      <div
        className="settings-definition-left"
      >
        <label
          className="h3"
          htmlFor="name"
        >
          settings.pr_decoration.binding.form.name
          <MandatoryFieldMarker
            className="spacer-right"
          />
        </label>
        <div
          className="markdown small spacer-top"
        >
          settings.pr_decoration.binding.form.name.help
        </div>
      </div>
      <div
        className="settings-definition-right"
      >
        <Select
          autosize={true}
          className="abs-width-400 big-spacer-top"
          clearable={false}
          id="name"
          menuContainerStyle={
            Object {
              "maxWidth": "210%",
              "width": "auto",
            }
          }
          onChange={[Function]}
          optionRenderer={[Function]}
          options={
            Array [
              Object {
                "alm": "github",
                "key": "i1",
                "url": "http://github.enterprise.com",
              },
              Object {
                "alm": "github",
                "key": "i2",
                "url": "http://github.enterprise.com",
              },
              Object {
                "alm": "bitbucket",
                "key": "i3",
                "url": "http://bbs.enterprise.com",
              },
              Object {
                "alm": "azure",
                "key": "i4",
              },
            ]
          }
          searchable={false}
          value=""
          valueKey="key"
          valueRenderer={[Function]}
        />
      </div>
    </div>
    <div
      className="display-flex-center big-spacer-top action-section"
    >
      <Button
        className="spacer-right"
        onClick={[MockFunction]}
      >
        <span
          data-test="project-settings__alm-reset"
        >
          reset_verb
        </span>
      </Button>
      <Button
        disabled={false}
        onClick={[MockFunction]}
      >
        settings.pr_decoration.binding.check_configuration
        <DeferredSpinner
          className="spacer-left"
          loading={false}
        />
      </Button>
    </div>
    <Alert
      className="big-spacer-top"
      display="inline"
      variant="error"
    >
      <p
        className="spacer-bottom"
      >
        settings.pr_decoration.binding.check_configuration.failure
      </p>
      <ul
        className="list-styled"
      >
        <li
          key="0"
        >
          Test
        </li>
        <li
          key="1"
        >
          tesT
        </li>
      </ul>
      <p>
        settings.pr_decoration.binding.check_configuration.contact_admin
      </p>
    </Alert>
  </form>
</div>
`;

exports[`should render correctly: with a single ALM instance 1`] = `
<div>
  <header
    className="page-header"
  >
    <h1
      className="page-title"
    >
      settings.pr_decoration.binding.title
    </h1>
  </header>
  <div
    className="markdown small spacer-top big-spacer-bottom"
  >
    settings.pr_decoration.binding.description
  </div>
  <form
    onSubmit={[Function]}
  >
    <MandatoryFieldsExplanation
      className="form-field"
    />
    <div
      className="settings-definition big-spacer-bottom"
    >
      <div
        className="settings-definition-left"
      >
        <label
          className="h3"
          htmlFor="name"
        >
          settings.pr_decoration.binding.form.name
          <MandatoryFieldMarker
            className="spacer-right"
          />
        </label>
        <div
          className="markdown small spacer-top"
        >
          settings.pr_decoration.binding.form.name.help
        </div>
      </div>
      <div
        className="settings-definition-right"
      >
        <Select
          autosize={true}
          className="abs-width-400 big-spacer-top"
          clearable={false}
          id="name"
          menuContainerStyle={
            Object {
              "maxWidth": "210%",
              "width": "auto",
            }
          }
          onChange={[Function]}
          optionRenderer={[Function]}
          options={
            Array [
              Object {
                "alm": "github",
                "key": "i1",
                "url": "http://github.enterprise.com",
              },
            ]
          }
          searchable={false}
          value=""
          valueKey="key"
          valueRenderer={[Function]}
        />
      </div>
    </div>
    <div
      className="display-flex-center big-spacer-top action-section"
    />
  </form>
</div>
`;

exports[`should render correctly: with a valid and saved form 1`] = `
<div>
  <header
    className="page-header"
  >
    <h1
      className="page-title"
    >
      settings.pr_decoration.binding.title
    </h1>
  </header>
  <div
    className="markdown small spacer-top big-spacer-bottom"
  >
    settings.pr_decoration.binding.description
  </div>
  <form
    onSubmit={[Function]}
  >
    <MandatoryFieldsExplanation
      className="form-field"
    />
    <div
      className="settings-definition big-spacer-bottom"
    >
      <div
        className="settings-definition-left"
      >
        <label
          className="h3"
          htmlFor="name"
        >
          settings.pr_decoration.binding.form.name
          <MandatoryFieldMarker
            className="spacer-right"
          />
        </label>
        <div
          className="markdown small spacer-top"
        >
          settings.pr_decoration.binding.form.name.help
        </div>
      </div>
      <div
        className="settings-definition-right"
      >
        <Select
          autosize={true}
          className="abs-width-400 big-spacer-top"
          clearable={false}
          id="name"
          menuContainerStyle={
            Object {
              "maxWidth": "210%",
              "width": "auto",
            }
          }
          onChange={[Function]}
          optionRenderer={[Function]}
          options={
            Array [
              Object {
                "alm": "github",
                "key": "i1",
                "url": "http://github.enterprise.com",
              },
              Object {
                "alm": "github",
                "key": "i2",
                "url": "http://github.enterprise.com",
              },
              Object {
                "alm": "bitbucket",
                "key": "i3",
                "url": "http://bbs.enterprise.com",
              },
              Object {
                "alm": "azure",
                "key": "i4",
              },
            ]
          }
          searchable={false}
          value="i1"
          valueKey="key"
          valueRenderer={[Function]}
        />
      </div>
    </div>
    <AlmSpecificForm
      alm="github"
      formData={
        Object {
          "key": "i1",
          "monorepo": false,
          "repository": "account/repo",
        }
      }
      instances={
        Array [
          Object {
            "alm": "github",
            "key": "i1",
            "url": "http://github.enterprise.com",
          },
          Object {
            "alm": "github",
            "key": "i2",
            "url": "http://github.enterprise.com",
          },
          Object {
            "alm": "bitbucket",
            "key": "i3",
            "url": "http://bbs.enterprise.com",
          },
          Object {
            "alm": "azure",
            "key": "i4",
          },
        ]
      }
      monorepoEnabled={false}
      onFieldChange={[MockFunction]}
    />
    <div
      className="display-flex-center big-spacer-top action-section"
    >
      <Button
        className="spacer-right"
        onClick={[MockFunction]}
      >
        <span
          data-test="project-settings__alm-reset"
        >
          reset_verb
        </span>
      </Button>
      <Button
        disabled={false}
        onClick={[MockFunction]}
      >
        settings.pr_decoration.binding.check_configuration
        <DeferredSpinner
          className="spacer-left"
          loading={false}
        />
      </Button>
    </div>
    <Alert
      className="big-spacer-top"
      display="inline"
      variant="success"
    >
      settings.pr_decoration.binding.check_configuration.success
    </Alert>
  </form>
</div>
`;

exports[`should render correctly: with an empty form 1`] = `
<div>
  <header
    className="page-header"
  >
    <h1
      className="page-title"
    >
      settings.pr_decoration.binding.title
    </h1>
  </header>
  <div
    className="markdown small spacer-top big-spacer-bottom"
  >
    settings.pr_decoration.binding.description
  </div>
  <form
    onSubmit={[Function]}
  >
    <MandatoryFieldsExplanation
      className="form-field"
    />
    <div
      className="settings-definition big-spacer-bottom"
    >
      <div
        className="settings-definition-left"
      >
        <label
          className="h3"
          htmlFor="name"
        >
          settings.pr_decoration.binding.form.name
          <MandatoryFieldMarker
            className="spacer-right"
          />
        </label>
        <div
          className="markdown small spacer-top"
        >
          settings.pr_decoration.binding.form.name.help
        </div>
      </div>
      <div
        className="settings-definition-right"
      >
        <Select
          autosize={true}
          className="abs-width-400 big-spacer-top"
          clearable={false}
          id="name"
          menuContainerStyle={
            Object {
              "maxWidth": "210%",
              "width": "auto",
            }
          }
          onChange={[Function]}
          optionRenderer={[Function]}
          options={
            Array [
              Object {
                "alm": "github",
                "key": "i1",
                "url": "http://github.enterprise.com",
              },
              Object {
                "alm": "github",
                "key": "i2",
                "url": "http://github.enterprise.com",
              },
              Object {
                "alm": "bitbucket",
                "key": "i3",
                "url": "http://bbs.enterprise.com",
              },
              Object {
                "alm": "azure",
                "key": "i4",
              },
            ]
          }
          searchable={false}
          value=""
          valueKey="key"
          valueRenderer={[Function]}
        />
      </div>
    </div>
    <div
      className="display-flex-center big-spacer-top action-section"
    />
  </form>
</div>
`;

exports[`should render correctly: with no ALM instances (admin user) 1`] = `
<div>
  <Alert
    className="spacer-top huge-spacer-bottom"
    variant="info"
  >
    <FormattedMessage
      defaultMessage="settings.pr_decoration.binding.no_bindings.admin"
      id="settings.pr_decoration.binding.no_bindings.admin"
      values={
        Object {
          "link": <Link
            onlyActiveOnIndex={false}
            style={Object {}}
            to={
              Object {
                "pathname": "/admin/settings",
                "query": Object {
                  "category": "almintegration",
                },
              }
            }
          >
            settings.pr_decoration.binding.no_bindings.link
          </Link>,
        }
      }
    />
  </Alert>
</div>
`;

exports[`should render correctly: with no ALM instances (non-admin user) 1`] = `
<div>
  <Alert
    className="spacer-top huge-spacer-bottom"
    variant="info"
  >
    settings.pr_decoration.binding.no_bindings
  </Alert>
</div>
`;

exports[`should render select options correctly 1`] = `
<React.Fragment>
  <span>
    i1
     — 
  </span>
  <span
    className="text-muted"
  >
    http://github.enterprise.com
  </span>
</React.Fragment>
`;

exports[`should render select options correctly 2`] = `
<React.Fragment>
  <span>
    i2
     — 
  </span>
  <span
    className="text-muted"
  >
    http://github.enterprise.com
  </span>
</React.Fragment>
`;
